api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,231 @@
+package legacysync
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/queue"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/harmony-one/harmony/consensus"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/node/worker"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/shard"
+	"github.com/pkg/errors"
+)
+
+type EpochSync struct {
+	beaconChain        blockChain
+	selfip             string
+	selfport           string
+	selfPeerHash       [20]byte // hash of ip and address combination
+	commonBlocks       map[int]*types.Block
+	lastMileBlocks     []*types.Block // last mile blocks to catch up with the consensus
+	syncConfig         *SyncConfig
+	isExplorer         bool
+	stateSyncTaskQueue *queue.Queue
+	syncMux            sync.Mutex
+	lastMileMux        sync.Mutex
+
+	syncStatus syncStatus
+}
+
+// GetSyncStatus get the last sync status for other modules (E.g. RPC, explorer).
+// If the last sync result is not expired, return the sync result immediately.
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *EpochSync) GetSyncStatus() SyncCheckResult {
+	return ss.syncStatus.Get(func() SyncCheckResult {
+		return ss.isInSync(false)
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(_ bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	if otherHeight1 == math.MaxUint64 {
+		utils.Logger().Error().
+			Uint64("OtherHeight", otherHeight1).
+			Int("Peers count", ss.syncConfig.PeersCount()).
+			Msg("[EPOCHSYNC] No peers for get height")
+		return SyncCheckResult{}
+	}
+	curEpoch := ss.beaconChain.CurrentBlock().Epoch().Uint64()
+	otherEpoch := shard.Schedule.CalcEpochNumber(otherHeight1).Uint64()
+	normalizedOtherEpoch := otherEpoch - 1
+	inSync := curEpoch == normalizedOtherEpoch
+
+	heightDiff := normalizedOtherEpoch - curEpoch
+	if normalizedOtherEpoch < curEpoch {
+		heightDiff = 0
+	}
+
+	utils.Logger().Info().
+		Uint64("OtherEpoch", otherEpoch).
+		Uint64("CurrentEpoch", curEpoch).
+		Msg("[EPOCHSYNC] Checking sync status")
+	return SyncCheckResult{
+		IsInSync:    inSync,
+		OtherHeight: otherHeight1,
+		HeightDiff:  heightDiff,
```

Comment:
EpochDiff.

Reply:
SyncCheckResult is struct with field HeightDiff, and it used in other places too, for example in legasy sync, where 
HeightDiff is appropriate name. I can't just rename it, it will lead to other changes. Suggest leave it as is.